File: licensing/storage/postgres/jti_ledger.go
package postgres

import (
	"context"
	"fmt"

	"github.com/jackc/pgx/v5/pgxpool"

	"github.com/AnoRebel/licensing/licensing/client"
)

// JtiLedger is the Postgres-backed replay-prevention ledger. It
// implements client.JtiLedger using the jti_uses table created by
// migration 0003.
//
// Concurrency: pgxpool.Pool serializes per-connection access; the
// ledger is safe for concurrent use from any number of goroutines.
// Each RecordJtiUse is one INSERT round trip.
type JtiLedger struct {
	pool *pgxpool.Pool
}

// NewJtiLedger constructs a Postgres-backed jti ledger. The caller
// MUST have applied migrations (jti_uses lives in 0003_jti_uses.sql)
// before the first RecordJtiUse call.
func NewJtiLedger(pool *pgxpool.Pool) *JtiLedger {
	return &JtiLedger{pool: pool}
}

// RecordJtiUse implements client.JtiLedger.
//
// Strategy: INSERT ... ON CONFLICT DO NOTHING. The CommandTag's
// RowsAffected returns 0 on conflict, 1 on fresh insert, distinguishing
// first-use from replay without a separate SELECT.
func (l *JtiLedger) RecordJtiUse(jti string, expSec int64) (bool, error) {
	tag, err := l.pool.Exec(
		context.Background(),
		`INSERT INTO jti_uses (jti, expires_at) VALUES ($1, $2)
		 ON CONFLICT (jti) DO NOTHING`,
		jti, expSec,
	)
	if err != nil {
		return false, fmt.Errorf("record jti use: %w", err)
	}
	return tag.RowsAffected() == 1, nil
}

// PruneExpired implements client.JtiLedger. Removes every row whose
// expires_at ≤ nowSec. Idempotent.
func (l *JtiLedger) PruneExpired(nowSec int64) (int, error) {
	tag, err := l.pool.Exec(
		context.Background(),
		`DELETE FROM jti_uses WHERE expires_at <= $1`,
		nowSec,
	)
	if err != nil {
		return 0, fmt.Errorf("prune jti_uses: %w", err)
	}
	return int(tag.RowsAffected()), nil
}

// Compile-time interface check.
var _ client.JtiLedger = (*JtiLedger)(nil)

File: typescript/src/storage/postgres/jti-ledger.ts
/**
 * Postgres-backed JtiLedger adapter.
 *
 * Implements `client.JtiLedger` against the `jti_uses` table created by
 * migration `0003_jti_uses.sql`. The pg `Pool` serializes per-connection
 * access; the ledger is safe to share across async call sites. Each
 * `recordJtiUse` is one INSERT round trip.
 *
 * Operators using the high-level `PostgresStorage` already have the
 * underlying `Pool`; pull it out and pass to `new PostgresJtiLedger(pool)`.
 * The migration MUST be applied before the first `recordJtiUse`.
 */

import type { Pool } from 'pg';

import type { JtiLedger } from '../../client/jti-ledger.ts';

export class PostgresJtiLedger implements JtiLedger {
  readonly #pool: Pool;

  constructor(pool: Pool) {
    this.#pool = pool;
  }

  /**
   * Strategy: INSERT ... ON CONFLICT DO NOTHING. The pg result's
   * `rowCount` reflects whether the insert wrote a row — 1 on fresh
   * insert, 0 on conflict. Distinguishes first-use from replay without
   * a separate SELECT.
   */
  async recordJtiUse(jti: string, expSec: number): Promise<boolean> {
    const result = await this.#pool.query(
      'INSERT INTO jti_uses (jti, expires_at) VALUES ($1, $2) ON CONFLICT (jti) DO NOTHING',
      [jti, expSec],
    );
    return result.rowCount === 1;
  }

  /**
   * Removes every row whose `expires_at ≤ nowSec`. Idempotent; returns
   * the number of rows removed.
   */
  async pruneExpired(nowSec: number): Promise<number> {
    const result = await this.#pool.query('DELETE FROM jti_uses WHERE expires_at <= $1', [nowSec]);
    return result.rowCount ?? 0;
  }
}
